clang/test/Driver/sycl-intel-gpu.cpp

```diff
@@ -0,0 +1,146 @@
+// RUN:   FileCheck %s --check-prefix=BAD_INPUT
+// RUN: %clang_cl -c -fsycl -fsycl-targets=intel_gpu_bad -### %s 2>&1 | \
+// RUN:   FileCheck %s --check-prefix=BAD_INPUT
+// BAD_INPUT: error: SYCL target is invalid: 'intel_gpu_bad'
+
+/// Test for proper creation of fat object
+// RUN: %clangxx -c -fsycl -fsycl-targets=intel_gpu_skl \
+// RUN:   -target x86_64-unknown-linux-gnu -### %s 2>&1 | \
+// RUN:   FileCheck %s --check-prefix=FATO
+// FATO: clang-offload-bundler{{.*}} "-type=o"
+// FATO: "-targets=sycl-spir64_gen-unknown-unknown-skl,host-x86_64-unknown-linux-gnu"
+
+/// Test for proper consumption of fat object
+// RUN: touch %t.o
+// RUN: %clangxx -fsycl -fsycl-targets=intel_gpu_skl \
+// RUN:   -target x86_64-unknown-linux-gnu -### %t.o 2>&1 | \
+// RUN:   FileCheck %s --check-prefix=CONSUME_FAT
+// CONSUME_FAT: clang-offload-bundler{{.*}} "-type=o"
+// CONSUME_FAT: "-targets=host-x86_64-unknown-linux-gnu,sycl-spir64_gen-unknown-unknown-skl"
+// CONSUME_FAT: "-unbundle" "-allow-missing-bundles"
+
+/// Test phases, BoundArch settings used for -device target. Additional
+/// offload action used for compilation and backend compilation.
+// RUN: %clangxx -fsycl -fsycl-targets=intel_gpu_skl -fno-sycl-device-lib=all \
+// RUN:   -fno-sycl-instrument-device-code \
+// RUN:   -target x86_64-unknown-linux-gnu -ccc-print-phases %s 2>&1 | \
+// RUN:   FileCheck %s --check-prefix=CHECK_PHASES
+// CHECK_PHASES: 0: input, "[[INPUT:.+\.cpp]]", c++, (host-sycl)
+// CHECK_PHASES: 1: append-footer, {0}, c++, (host-sycl)
+// CHECK_PHASES: 2: preprocessor, {1}, c++-cpp-output, (host-sycl)
+// CHECK_PHASES: 3: input, "[[INPUT]]", c++, (device-sycl, skl)
+// CHECK_PHASES: 4: preprocessor, {3}, c++-cpp-output, (device-sycl, skl)
+// CHECK_PHASES: 5: compiler, {4}, ir, (device-sycl, skl)
+// CHECK_PHASES: 6: offload, "device-sycl (spir64_gen-unknown-unknown:skl)" {5}, ir
+// CHECK_PHASES: 7: offload, "host-sycl (x86_64-unknown-linux-gnu)" {2}, "device-sycl (spir64_gen-unknown-unknown:skl)" {6}, c++-cpp-output
+// CHECK_PHASES: 8: compiler, {7}, ir, (host-sycl)
+// CHECK_PHASES: 9: backend, {8}, assembler, (host-sycl)
+// CHECK_PHASES: 10: assembler, {9}, object, (host-sycl)
+// CHECK_PHASES: 11: linker, {10}, image, (host-sycl)
+// CHECK_PHASES: 12: linker, {6}, ir, (device-sycl)
+// CHECK_PHASES: 13: sycl-post-link, {12}, tempfiletable, (device-sycl)
+// CHECK_PHASES: 14: file-table-tform, {13}, tempfilelist, (device-sycl, skl)
+// CHECK_PHASES: 15: llvm-spirv, {14}, tempfilelist, (device-sycl, skl)
+// CHECK_PHASES: 16: backend-compiler, {15}, image, (device-sycl, skl)
+// CHECK_PHASES: 17: offload, "device-sycl (spir64_gen-unknown-unknown:skl)" {16}, image
+// CHECK_PHASES: 18: file-table-tform, {13, 17}, tempfiletable, (device-sycl)
+// CHECK_PHASES: 19: clang-offload-wrapper, {18}, object, (device-sycl)
+// CHECK_PHASES: 20: offload, "host-sycl (x86_64-unknown-linux-gnu)" {11}, "device-sycl (spir64_gen-unknown-unknown)" {19}, image
```

Comment:
After reading all the patch, I think the root of all the issues and workarounds simply boils down to setting correctly `20` (having `"device-sycl (spir64_gen-unknown-unknown:skl)" {19}, image` instead of `"device-sycl (spir64_gen-unknown-unknown)" {19}, image`

Reply:
Thanks for the feedback - I have cleaned up the dependency settings to allow for propagation up as expected.